src/EFCore.Relational/Metadata/Conventions/TableSharingConcurrencyTokenConvention.cs

```diff
@@ -0,0 +1,198 @@
+                                || declaringEntityType.IsInOwnershipPath(entityType)
+                                || entityType.IsInOwnershipPath(declaringEntityType))
+                            {
+                                foundMappedProperty = true;
+                                break;
+                            }
+                        }
+
+                        foundMappedProperty = foundMappedProperty
+                            || entityType.GetAllBaseTypes().SelectMany(t => t.GetDeclaredProperties())
+                                .Any(p => p.GetColumnName() == concurrencyColumnName);
+
+                        if (!foundMappedProperty)
+                        {
+                            // store the entity type which is missing the
+                            // concurrency token property, mapped to an example
+                            // property which _is_ mapped to this concurrency token
+                            // column and which will be used later as a template
+                            entityTypesMissingConcurrencyColumn.Add(
+                                entityType, propertiesMappedToConcurrencyColumn.First());
+                        }
+                    }
+
+                    RemoveDerivedEntityTypes(ref entityTypesMissingConcurrencyColumn);
+
+                    foreach(var entityTypeToExampleProperty in entityTypesMissingConcurrencyColumn)
+                    {
+                        var entityType = entityTypeToExampleProperty.Key;
+                        var exampleProperty = entityTypeToExampleProperty.Value;
+                        var concurrencyShadowPropertyBuilder =
+#pragma warning disable EF1001 // Internal EF Core API usage.
+                            ((InternalEntityTypeBuilder)entityType.Builder).CreateUniqueProperty(
+                                ConcurrencyPropertyPrefix + exampleProperty.Name,
+                                exampleProperty.ClrType,
+                                !exampleProperty.IsNullable);
+                        concurrencyShadowPropertyBuilder.SetColumnName(concurrencyColumnName);
+                        _ = concurrencyShadowPropertyBuilder.SetIsConcurrencyToken(true, ConfigurationSource.Convention);
+                        _ = concurrencyShadowPropertyBuilder.SetValueGenerated(exampleProperty.ValueGenerated, ConfigurationSource.Convention);
+#pragma warning restore EF1001 // Internal EF Core API usage.
+                    }
+                }
+            }
+        }
+
+        private void GetMappings(IConventionModel model,
+            out Dictionary<Tuple<string, string>, IList<IConventionEntityType>> tableToEntityTypes,
+            out Dictionary<Tuple<string, string>, Dictionary<string, IList<IConventionProperty>>> concurrencyColumnsToProperties)
+        {
+            tableToEntityTypes = new Dictionary<Tuple<string, string>, IList<IConventionEntityType>>();
```

Comment:
Use value tuples instead
```suggestion
            tableToEntityTypes = new Dictionary<(string Table, string Schema), IList<IConventionEntityType>>();
```